feat: generate and persist application storage unique ID

adisazhar123 opened the pull request:
When we create a k8s deployment, a unique identifier which we call the storage unique ID is used to construct the PVC name. This ID is attached as an annotation in the statefulset, daemonset, and service.

This lays the groundwork for https://github.com/juju/juju/pull/21309. When we re-create a statefulset for a storage update, we need the storage unique ID so that it's consistent across the lifecycle of the deployment. Thus, by persisting in mongo, we can easily grab the ID rather than generating it everytime a statefulset is created (this is what the existing implementation does). Without persisting the storage unique ID, we risk **losing** it because in an unfortunate circumstance where the statefulset is orphan deleted and re-applied, we won't be able to get the original storage unique ID.

What's in this PR:

- create a new `storageUniqueID` field in the application document
- the `storageUniqueID` is generated and persisted when the application is created
- wire up `storageUniqueID` in ProvisioningInfo facade method
- let k8s client read `storageUniqueID` returned from the facade to populate the annotation and PVC names
- upgrade step to backfill the missing `storageUniqueID` for existing applications
- migration to populate missing `storageUniqueID` for applications in an old model => this relies on https://github.com/juju/description/pull/188


Depends on https://github.com/juju/description/pull/188. The tests should pass once PR 188 is merged.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing
- ~[ ] [Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing~
- ~[ ] [doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages~

## QA steps

#### Storage unique ID for new applications

Save image to minikube / microk8s (replace the name)
```
make minikube-operator-update
```

Bootstrap a k8s controller.
```
juju bootstrap minikube adisctrl
```
Add model.
```
juju add-model m1
```
Deploy a charm
```
juju deploy postgresql-k8s --channel=14/stable --trust
```

Check mongodb for the storage-unique-id field (follow [this guide](https://discourse.charmhub.io/t/login-into-mongodb/309) to access the mongo instance in ks8)

```
db.applications.find().pretty()
```
<img width="1269" height="200" alt="image" src="https://github.com/user-attachments/assets/f192fe6b-784b-4ae8-82da-70b6e0c0bea3" />

---

#### Populate the storage unique ID for existing applications by performing an upgrade

Existing applications won't have a storage unique ID saved in the application's document. Juju needs to undergo an upgrade step so that we can populate the IDs.


In this case I'm using 3.6.12. I have the `jujud-operator:3.6.12` pushed to my docker repo.
You can do this by running 
```
docker pull ghcr.io/juju/jujud-operator:3.6.12
docker tag ghcr.io/juju/jujud-operator:3.6.12 adisazhar123/jujud-operator:3.6.12
docker push adisazhar123/jujud-operator:3.6.12
```

Make sure you also have `juju-db` and `charm-base` in your docker repo. 

```
OCI_REGISTRY_USERNAME=docker.io/adisazhar123 JUJU_DB_VERSION=4.4 make seed-repository
```

Bootstrap an older version.

```
juju bootstrap minikube adisctrl --config caas-image-repo=docker.io/adisazhar123 --agent-version 3.6.12
```


Add model
```
juju add-model m1
```

Deploy app. This won't have the storage unique ID saved in the application document. It's saved in the statefulset annotation.
```
juju deploy postgresql-k8s --channel=14/stable --trust
```

<img width="1269" height="169" alt="image" src="https://github.com/user-attachments/assets/202ca04b-1689-45ad-a9f8-2261078bbd1a" />


Checkout the tip of this branch. Then build and push the jujud-operator image.

```
JUJU_BUILD_NUMBER=1 OCI_REGISTRY_USERNAME=docker.io/adisazhar123 make push-operator-image
JUJU_BUILD_NUMBER=1 OCI_REGISTRY_USERNAME=docker.io/adisazhar123 make minikube-operator-update
```

Now perform the upgrade. 

```
juju upgrade-controller --agent-version=3.6.13.1
```

The storage unique ID field should be populated now. Check it in mongo.
```
db.applications.find().pretty()
```
<img width="1269" height="169" alt="image" src="https://github.com/user-attachments/assets/0b392886-61e2-4ba7-ac1f-3bacf26ce770" />

It has the same value as the statefulset annotation.

<img width="1099" height="249" alt="image" src="https://github.com/user-attachments/assets/22ff91e4-5b74-4cc8-b2b2-37a0071b2053" />


---

#### Populate the storage unique ID when migrating from an old model to a >= 3.6.13 controller

Bootstrap a 3.6.9 controller

```
juju bootstrap minikube source --agent-version 3.6.9
```

Add a model that we want to migrate... and deploy an app

```
juju add-model model1
juju deploy postgresql-k8s --channel=14/stable --trust
```

Check the storage unique ID in mongo is non-existent

```
db.applications.find({"name": "postgresql-k8s"}, {"name": 1, "storage-unique-id": 1}).pretty()
```

<img width="1799" height="197" alt="image" src="https://github.com/user-attachments/assets/13b09415-6e05-4823-992d-35c4718ca194" />


Bootstrap a 3.6.13 controller. This is the controller we want to migrate to.

```
make minikube-operator-update
juju bootstrap minikube target 
```

Switch to source:model1 to perform the migration

```
juju switch source:model1
juju migrate model1 target
```

<img width="1100" height="165" alt="image" src="https://github.com/user-attachments/assets/15d8b411-0e47-4b85-aace-0b246b3a2013" />


Switch to target:model1 to check that it's migrated

```
juju switch target:model1
```

<img width="1710" height="378" alt="image" src="https://github.com/user-attachments/assets/af12107e-1367-44d1-8809-c60f4a03668d" />


Check that the storage unique ID is populated in mongo

```
db.applications.find({"name": "postgresql-k8s"}, {"name": 1, "storage-unique-id": 1}).pretty()
```

<img width="1765" height="239" alt="image" src="https://github.com/user-attachments/assets/975b8e59-4c47-4054-8ec3-852d65f03616" />


## Documentation changes

N/A

## Links


**Issue:** Fixes https://github.com/juju/juju/issues/19829.

**Jira card:** [JUJU-8589](https://warthogs.atlassian.net/browse/JUJU-8589)


[JUJU-8589]: https://warthogs.atlassian.net/browse/JUJU-8589?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

adisazhar123 committed:
feat: generate and persist storageuniqueid in the state layer

adisazhar123 committed:
feat: wire up storageUniqueID in facades

adisazhar123 committed:
feat: use storageUniqueID from ProvisioningInfo facade call

adisazhar123 committed:
feat: add upgrade step to populate storage unique id from annotations

adisazhar123 committed:
test: troubleshooting upgrades

adisazhar123 committed:
feat: populate storageuniqueid for each model

adisazhar123 committed:
feat: remove randomPrefixFunc

adisazhar123 committed:
test: pass storageUniqueId instead of generating it

adisazhar123 committed:
fix: errors from static-analysis

adisazhar123 committed:
fix: errors from static-analysis

wallyworld commented on api/client/modelupgrader/upgrader.go:
```diff
@@ -60,6 +61,7 @@ func (c *Client) UpgradeModel(
 		IgnoreAgentVersions: ignoreAgentVersions,
 		DryRun:              druRun,
 	}
+	log.Printf("[adis] upgrade model. args: %+v\n", args)
```
delete me

wallyworld commented on apiserver/facades/client/modelupgrader/upgrader.go:
```diff
@@ -129,6 +129,7 @@ func (m *ModelUpgraderAPI) AbortModelUpgrade(arg params.ModelParam) error {
 // UpgradeModel upgrades a model.
 func (m *ModelUpgraderAPI) UpgradeModel(arg params.UpgradeModelParams) (result params.UpgradeModelResult, err error) {
 	logger.Tracef("UpgradeModel arg %#v", arg)
+	logger.Infof("[adis] here in upgrade model server with arg: %+v", arg)
```
delete me

wallyworld commented on apiserver/facades/controller/caasapplicationprovisioner/provisioner.go:
```diff
@@ -435,6 +435,7 @@ func (a *API) provisioningInfo(appName names.ApplicationTag) (*params.CAASApplic
 		CharmURL:             *charmURL,
 		Trust:                appConfig.GetBool(application.TrustConfigOptionName, false),
 		Scale:                app.GetScale(),
+		StorageUniqueID:      app.StorageUniqueID(),
```
Should have mentioned this before, we should use `GetStorageUniqueID()`

wallyworld commented on caas/broker.go:
```diff
@@ -348,6 +352,10 @@ type ServiceManager interface {
+
+	// LabelVersion returns the detected label version for k8s resources created
+	// for this model.
+	LabelVersion() constants.LabelVersion
```
This isn't part of this PR?

wallyworld commented on internal/provider/kubernetes/admissionregistration_test.go:
```diff
@@ -764,8 +761,9 @@ func (s *K8sBrokerSuite) assertValidatingWebhookConfigurations(c *gc.C, cfgs []k
-		ResourceTags: map[string]string{"juju-controller-uuid": testing.ControllerTag.Id()},
+		ImageDetails:    resources.DockerImageDetails{RegistryPath: "operator/image-path"},
+		ResourceTags:    map[string]string{"juju-controller-uuid": testing.ControllerTag.Id()},
+		StorageUniqueID: "appuuid",
```
To avoid confusion, maybe use "uniqueid" or something (like in other tests)

wallyworld commented on internal/provider/kubernetes/k8s.go:
```diff
@@ -725,6 +720,7 @@ func (k *kubernetesClient) GetService(appName string, mode caas.DeploymentMode,
 		result.Status = status.StatusInfo{
 			Status:  ssStatus,
 			Message: message,
+			Data:    toInterfaceMap(ss.Annotations),
```
What's the reason for this change?

wallyworld commented on internal/worker/caasapplicationprovisioner/ops.go:
```diff
@@ -133,6 +133,7 @@ func appAlive(appName string, app caas.Application, password string, lastApplied
 	if err != nil {
 		return errors.Annotate(err, "retrieving provisioning info")
 	}
+	logger.Infof("[adis] provisionInfo: %+v", provisionInfo)
```
delete me

adisazhar123 replied:
Was used for local testing. Done.

wallyworld commented on rpc/params/caas.go:
```diff
@@ -60,6 +60,7 @@ type CAASApplicationProvisioningInfo struct {
 	Trust                bool                         `json:"trust,omitempty"`
 	Scale                int                          `json:"scale,omitempty"`
 	Error                *Error                       `json:"error,omitempty"`
+	StorageUniqueID      string                       `json:"storage-unique-id,omitempty"`
```
This will never be empty now right? 
We should keep Error as the last attribute, so move this up higher.

wallyworld commented on rpc/params/kubernetes.go:
```diff
@@ -27,6 +27,7 @@ type KubernetesProvisioningInfo struct {
 	Devices              []KubernetesDeviceParams     `json:"devices,omitempty"`
 	CharmModifiedVersion int                          `json:"charm-modified-version,omitempty"`
 	ImageRepo            DockerImageInfo              `json:"image-repo,omitempty"`
+	StorageUniqueID      string                       `json:"storage-unique-id,omitempty"`
```
This will always be set right?

adisazhar123 replied:
`omitempty` removed. Done.

wallyworld commented on state/application.go:
```diff
@@ -4276,3 +4277,7 @@ func (st *State) WatchApplicationsWithPendingCharms() StringsWatcher {
 	})
 }
+
+func (a *Application) StorageUniqueID() string {
```
Comment me

wallyworld replied:
Needs a test

wallyworld commented on internal/worker/upgradesteps/worker.go:
```diff
@@ -318,7 +318,6 @@ func (w *upgradeSteps) runUpgradeSteps(agentConfig agent.ConfigSetter) error {
 	if err := w.entity.SetStatus(status.Started, fmt.Sprintf("upgrading to %v", w.toVersion), nil); err != nil {
 		return errors.Trace(err)
 	}
-
```
do we need this change?

adisazhar123 replied:
Nope. Will remove.

wallyworld commented on state/state.go:
```diff
@@ -1310,9 +1316,10 @@ func (st *State) AddApplication(args AddApplicationArgs) (_ *Application, err er
+		DesiredScale:    scale,
+		Placement:       placement,
+		HasResources:    hasResources,
+		StorageUniqueID: storageUniqueID,
```
Needs a test

adisazhar123 replied:
Yup, working on it.

wallyworld commented on upgrades/operations.go:
```diff
@@ -21,6 +21,7 @@ var stateUpgradeOperations = func() []Operation {
 	steps := []Operation{
 		upgradeToVersion{version.MustParse("3.6.4"), stateStepsFor364()},
 		upgradeToVersion{version.MustParse("3.6.5"), stateStepsFor365()},
+		upgradeToVersion{version.MustParse("3.6.11"), stateStepsFor3611()},
```
This will be 3.6.12

adisazhar123 replied:
Renamed to 3.6.12

wallyworld commented on upgrades/steps_3611.go:
```diff
@@ -0,0 +1,17 @@
+package upgrades
+
+// stateStepsFor3611 returns upgrade steps for Juju 3.6.11 that manipulate state directly.
+func stateStepsFor3611() []Step {
```
This will need a test update and rename everything to 3.6.12

wallyworld commented on upgrades/backend.go:
```diff
@@ -42,3 +51,7 @@ func (s stateBackend) AddVirtualHostKeys() error {
 	return state.SplitMigrationStatusMessages(s.pool)
 }
+
+func (s stateBackend) PopulateApplicationStorageUniqueID() error {
```
Comment me

adisazhar123 replied:
Done

wallyworld commented on state/upgrades.go:
```diff
@@ -250,3 +253,97 @@ func SplitMigrationStatusMessages(pool *StatePool) error {
 	return st.runRawTransaction(ops)
 }
+
+func PopulateApplicationStorageUniqueID(pool *StatePool, getBrokerFunc func(model *Model) (caas.Broker, error)) error {
```
Comment me and needs a test.

adisazhar123 replied:
Yup. Working on it.

adisazhar123 replied:
Done

wallyworld commented on state/upgrades.go:
```diff
@@ -250,3 +253,97 @@ func SplitMigrationStatusMessages(pool *StatePool) error {
 }
+
+func PopulateApplicationStorageUniqueID(pool *StatePool, getBrokerFunc func(model *Model) (caas.Broker, error)) error {
+	logger.Infof("[adis] inside PopulateApplicationStorageUniqueID")
```
delete these extra log lines

adisazhar123 replied:
Will delete them. I have them for my testing.

wallyworld commented on state/upgrades.go:
```diff
@@ -250,3 +253,97 @@ func SplitMigrationStatusMessages(pool *StatePool) error {
+				return err
+			}
+			var mode charm.DeploymentMode
+			if d := ch.Meta().Deployment; d != nil {
```
This won't be set for sidecar charms. so you can't reply on it.
There's a help function to get the charm metadata format `format := charm.MetaFormat(charmInfo.Charm())`
Then for sidecar charms you can just look for the statefulset.

adisazhar123 replied:
Done.

wallyworld commented on state/upgrades.go:
```diff
@@ -250,3 +253,97 @@ func SplitMigrationStatusMessages(pool *StatePool) error {
+				return map[string]interface{}{}, err
+			}
+			logger.Infof("[adis] getAnnotation data is: %+v", service.Status.Data)
+			return service.Status.Data, nil
```
????? We should not be messing with status data to get annotations off a k8s resource.

wallyworld commented on state/upgrades.go:
```diff
@@ -250,3 +253,97 @@ func SplitMigrationStatusMessages(pool *StatePool) error {
+		applications, closer := st.db().GetCollection(applicationsC)
+		defer closer()
+
+		iter := applications.Find(nil).Iter()
```
We can construct the query to filter out apps with storage id already set, then we don't need the extra check below.

adisazhar123 replied:
Good idea. Thank you.

wallyworld reviewed (commented):
A good start, thank you.
We need to add tests and not use status data to hold annotation info.

adisazhar123 committed:
refactor: method name to GetStorageUniqueID

adisazhar123 committed:
feat: construct closure to get storage id from annotation

adisazhar123 committed:
test: fix failing application test

adisazhar123 committed:
test: upgrade step to populate storage id test

wallyworld commented on internal/provider/kubernetes/k8s.go:
```diff
@@ -2737,3 +2704,11 @@ func isStateful(pod *core.Pod) bool {
 	return false
 }
+
+func toInterfaceMap(in map[string]string) map[string]interface{} {
```
delete me

wallyworld commented on state/upgrades.go:
```diff
@@ -250,3 +250,71 @@ func SplitMigrationStatusMessages(pool *StatePool) error {
+	}
+
+	var ops []txn.Op
+
```
Don't need these or the `st.runRawTransaction(ops)`
Just
```
return runForAllModelStates(pool, func(st *State) error {
...
}
```

wallyworld commented on upgrades/backend.go:
```diff
@@ -42,3 +55,129 @@ func (s stateBackend) AddVirtualHostKeys() error {
+
+func GetStorageUniqueID(newK8sClient newK8sFunc) func(appName string, model *state.Model) (string, error) {
+	k8sClientsByModel := make(map[string]kubernetesClient)
+	return func(appName string, model *state.Model) (string, error) {
```
Let's pass in a slice of app names for the model and return a slice of storage ids.
That way we only create the k8s client once per model.

wallyworld reviewed (commented):
Thanks for the progress, a few things to look at.

adisazhar123 committed:
fix: remove writing to data

adisazhar123 committed:
test: create a fake k8s impl

adisazhar123 committed:
fix: add func doc

adisazhar123 commented:
/build

adisazhar123 committed:
test: remove writing to data

this reverts to how it was previously

adisazhar123 commented:
### QAed after new change



**Deploying a new application shows storage-unique-id**

```
db.applications.find({}, {"name": 1, "storage-unique-id": 1}).pretty()
```

<img width="1269" height="200" alt="Screenshot from 2025-11-12 09-13-23" src="https://github.com/user-attachments/assets/0ba3d783-bca4-4215-ae43-b8b483b130be" />

**Storage unique ID for existing applications**

An existing application doesn't have storage-unique-id

```
db.applications.find({}, {"name": 1, "storage-unique-id": 1}).pretty()
```

<img width="1269" height="169" alt="image" src="https://github.com/user-attachments/assets/e0739cac-d728-4dfd-99f3-8d5a744c40fe" />

After an upgrade, the storage-unique-id is populated and see that it's indeed the same as the annotation value.

<img width="1269" height="169" alt="image" src="https://github.com/user-attachments/assets/2d02aa21-9156-4452-948f-956384a90e4e" />

```
kubectl get sts postgresql-k8s -n m1 -o json | jq '.metadata.annotations'
```

<img width="1099" height="249" alt="image" src="https://github.com/user-attachments/assets/4a998c39-b6c7-4312-a62b-68a0dd3eed50" />

adisazhar123 committed:
feat: change func signature to receive a slice of app names

This allows us to only create one k8s client for each model rather than using a cache to keep track what k8s client are opened for each model. It gets rid of that complexity.

adisazhar123 committed:
refactor: target version to 3.6.12

adisazhar123 committed:
test: adjust so that it receives a slice of app names

adisazhar123 committed:
test: target to version 3.6.12

adisazhar123 committed:
fix: remove logs used for testing

I kept the logs that may be beneficial to observe what happens during the upgrade

adisazhar123 commented on state/upgrades.go:
```diff
@@ -250,3 +250,71 @@ func SplitMigrationStatusMessages(pool *StatePool) error {
+				Update: bson.D{{"$set", bson.D{{"storage-unique-id", storageUniqueID}}}},
+			})
+		}
+		return nil
```
I have call it deferred. I think it's better that way in case an error occurs before we reach the return statement.

adisazhar123 committed:
fix: add func docs and remove unused func

adisazhar123 committed:
refactor: use `uniqueid` as the annotation value

adisazhar123 committed:
fix: remove `omitempty` from storageUniqueID field

for CAAS apps it's always non-empty so it makes sense to remove it

adisazhar123 committed:
test: assert that storageUniqueID has the corresponding value

for CAAS apps it's not empty because we use it to construct PVC names, whereas we don't need it for IAAS so we assert empty

adisazhar123 committed:
fix: static analysis errors and outdated mock

adisazhar123 commented on state/migration_internal_test.go:
```diff
@@ -413,6 +413,7 @@ func (s *MigrationSuite) TestApplicationDocFields(c *gc.C) {
 		"HasResources",
 		"ProvisioningState",
+		"StorageUniqueID",
 	)
```
@wallyworld I believe this requires changes in https://github.com/juju/description

I am working on it. It shouldn't block you from reviewing this PR.

adisazhar123 replied:
Ian, I thought about this more. We have a different storage unique ID mapping between 3.6 and 4.0.

If that's the case is the scope of `StorageUniqueID` migration only within 3.6 models?

adisazhar123 committed:
feat: add storage unique ID when exporting

adisazhar123 committed:
test: assert that storage unique ID exists in the export

adisazhar123 committed:
feat: add storage unique ID when importing

adisazhar123 committed:
test: assert that storage unique ID exists in the import

adisazhar123 committed:
feat: generate mock

adisazhar123 committed:
feat: populate missing storage unique IDs on model migration

adisazhar123 committed:
feat: update description package to v9.0.1

This version is aware of storage unique ID field

adisazhar123 committed:
feat: split into smaller funcs and introduce a Migrator interface

We were directly calling a concrete implementation of migration.ImportModel which makes it hard to test the storage unique ID migration changes (specific for CAAS). The Migrator interface will help us mock that implementation in unit tests. While the existing migration.ImportModel is still there for IAAS, at least this alleviates the testing for CAAS. Refactoring the migration.ImportModel for IAAS is for a future effort.

adisazhar123 committed:
feat: inject func to initialize k8sclient and migration wrapper

adisazhar123 committed:
feat: generator mock

adisazhar123 committed:
test: storage unique ID should be populated

This test case simulates a scenario where an exported a model has an app with a missing storage unique ID. The import determines to backfill the storage ID.

adisazhar123 committed:
feat: add go:generate to run mockgen

adisazhar123 committed:
test: take in a deserialized model and remove obsolete test case

adisazhar123 committed:
docs: add func docs

adisazhar123 committed:
refactor: remove testing logs

adisazhar123 commented on upgrades/steps_3612.go:
```diff
@@ -0,0 +1,17 @@
+package upgrades
+
+// stateStepsFor3612 returns upgrade steps for Juju 3.6.12 that manipulate state directly.
+func stateStepsFor3612() []Step {
```
@CodingCookieRookie @wallyworld [code of interest] this houses the upgrade steps when upgrading to 3.6.12. It invokes the logic to populate storage unique ID

adisazhar123 commented on internal/provider/kubernetes/application/application.go:
```diff
@@ -284,18 +277,13 @@ func (a *app) Ensure(config caas.ApplicationConfig) (err error) {
-		})
-		if err != nil {
-			return errors.Trace(err)
-		}
```
@CodingCookieRookie @wallyworld [code of interest] this is now removed in favor of `config.storageUniqueID`

adisazhar123 commented on migration/migration_test.go:
```diff
@@ -50,15 +49,6 @@ func (s *ImportSuite) SetUpTest(c *gc.C) {
-	c.Check(model, gc.IsNil)
-	c.Assert(err, gc.ErrorMatches, "yaml: unmarshal errors:\n.*")
-}
-
```
Not needed anymore since  migration.ImportModel expects a deserialized model

adisazhar123 commented on state/migration_export.go:
```diff
@@ -889,6 +889,7 @@ func (e *exporter) addApplication(ctx addApplicationContext) error {
 		MetricsCredentials:   application.doc.MetricCredentials,
 		PodSpec:              ctx.podSpecs[application.globalKey()],
+		StorageUniqueID:      application.doc.StorageUniqueID,
 	}
```
@CodingCookieRookie @wallyworld [code of interest] include storage unique ID in the exported model

adisazhar123 commented on state/migration_import.go:
```diff
@@ -1401,6 +1401,7 @@ func (i *importer) makeApplicationDoc(a description.Application) (*applicationDo
 		Placement:            a.Placement(),
 		HasResources:         a.HasResources(),
+		StorageUniqueID:      a.StorageUniqueID(),
 	}
```
@CodingCookieRookie @wallyworld [code of interest] make the import aware of storage unique ID so that on application creation it will include that field

wallyworld commented on upgrades/backend.go:
```diff
@@ -42,3 +56,175 @@ func (s stateBackend) AddVirtualHostKeys() error {
+		if err != nil {
+			return nil, err
+		}
+		k8sClient = kubernetesClient{k8sClient, cfg}
```
No need for the wrapper

wallyworld commented on upgrades/backend.go:
```diff
@@ -42,3 +56,175 @@ func (s stateBackend) AddVirtualHostKeys() error {
+// in the given model.
+// The storage unique ID is saved in annotations in a statefulset and for legacy
+// applications are saved in a deployment or daemonset.
+func GetStorageUniqueIDs(newK8sClient newK8sFunc) func(
```
Let's pass in a context here and have the caller use `context.Background)` if needed

adisazhar123 committed:
fix: static analysis errors

wallyworld commented on upgrades/backend.go:
```diff
@@ -42,3 +56,175 @@ func (s stateBackend) AddVirtualHostKeys() error {
+
+			// The app doesn't exist in deployment so we look at the daemonset
+			// resource.
+			daemonSet, err := k8sClient.AppsV1().
```
No need to do this if we found a statefulset above etc.
We could move these 3 chunks of code (for statefulset, deployment, daemonset) to separate functions and use a for loop to iterate them and stop as soon as we find the relevant artefact exists. Separately, we track whether that artefact has the storage id annotation so we know whether to generate it.

adisazhar123 replied:
I created a slice of funcs called `StorageUniqueIDFinder` in `kubernetes` package. Each func is an implementation to find the storage id from sts, deployment, and daemonset. 

The caller iterate through the slice and returns early if it finds one. Otherwise once it exhausts the loop, we indicate an error that a k8s resource isn't found for the given app.

We generate a storage id in `getUniqueIDFromAnnotation` func in `kubernetes` package if the annotation doesn't hold one. To the best of my knowledge, all apps except the controller app has a storage id.

wallyworld commented on upgrades/backend.go:
```diff
@@ -42,3 +56,175 @@ func (s stateBackend) AddVirtualHostKeys() error {
+}
+
+// NewK8sClient initializes a new k8s client for a given model.
+func NewK8sClient(model *state.Model) (kubernetes.Interface, *rest.Config, error) {
```
This doesn't need to be exported

adisazhar123 replied:
Done.

wallyworld commented on state/upgrades.go:
```diff
@@ -16,6 +16,12 @@ var _ = func() {
 	_ = applyToAllModelSettings(nil, nil)
 }
 
+type AppAndStorageID struct {
```
Comment me

wallyworld commented on state/state.go:
```diff
@@ -2666,3 +2673,12 @@ func TagFromDocID(docID string) names.Tag {
 }
+
+// RandomPrefix returns a random string.
+func RandomPrefix() (string, error) {
```
This doesn't really belong in state - it's an artefact of k8s storage.
I think we can move it to `storage/kubernetes.go`

wallyworld commented on state/upgrades.go:
```diff
@@ -250,3 +256,68 @@ func SplitMigrationStatusMessages(pool *StatePool) error {
+		logger.Debugf("trying to populate storage unique ID for apps in model %q", model.Name())
+
+		if model.Type() != ModelTypeCAAS {
+			logger.Debugf("skipping because model %q is not a caas model", model.Name())
```
`s/caas/k8s`

adisazhar123 replied:
Done

wallyworld commented on state/upgrades.go:
```diff
@@ -250,3 +256,68 @@ func SplitMigrationStatusMessages(pool *StatePool) error {
+			return nil
+		}
+
+		applications, closer := st.db().GetCollection(applicationsC)
```
We normally name this `applicationsColl`

adisazhar123 replied:
Done

wallyworld reviewed (commented):
Good progress, a few small things to start with.

wallyworld commented on state/upgrades_test.go:
```diff
@@ -207,3 +214,120 @@ func (s *upgradesSuite) TestSplitMigrationStatusMessages(c *gc.C) {
+
+	err = PopulateApplicationStorageUniqueID(s.pool, getStorageUniqueID)
+	c.Assert(err, gc.IsNil)
+	c.Assert(appMigratedCount, gc.Equals, 4)
```
We're not asserting the value of content of the storage id written to the database that I can see. We need to check any existing value (for app3) is not backfilled and the new ids get written as expected.

wallyworld commented on apiserver/facades/controller/migrationtarget/migrationtarget.go:
```diff
@@ -50,13 +77,46 @@ type APIV2 struct {
+type kubernetesClient struct {
+	kubernetes.Interface
+	config *rest.Config
+}
```
I don't think we need this wrapper

wallyworld commented on apiserver/facades/controller/migrationtarget/migrationtarget.go:
```diff
@@ -151,7 +213,59 @@ with an earlier version of the target controller and try again.
+
+	// We have this condition to handle backfilling storage unique IDs for
+	// caas deployments with a missing storage ID in the application doc.
+	if descriptionModel.Type() == "caas" {
```
There's a core model const for "caas"
Let's move this code to its own func

wallyworld commented on apiserver/facades/controller/migrationtarget/register.go:
```diff
@@ -82,5 +93,26 @@ func newFacade(ctx facade.Context, facadeVersions facades.FacadeVersions) (*API,
+
+// migrationWrapper implements [Migrator] and wraps around [migration.ImportModel]
+// for easy unit testing.
+type migrationWrapper struct{}
```
Don't need this.
Just define a func and pass that in
```
type ImportModelFunc func(
	importer migration.StateImporter,
	getClaimer migration.ClaimerFunc,
	model description.Model,
) (*state.Model, *state.State, error)
```

wallyworld reviewed (commented):
A few more small things

adisazhar123 committed:
feat: introduce a slice of func to hold storageid finder

The caller will iterate over this slice to find a storage id for a given application.

adisazhar123 committed:
refactor: move func to generate random string to k8s storage

adisazhar123 committed:
refactor: remove redundant baseModel and kubernetesClient

adisazhar123 committed:
refactor: use the slice of func to find storage id

adisazhar123 committed:
refactor: remove migrator mock

adisazhar123 committed:
feat: inject migration.ImportModel

adisazhar123 committed:
feat: add func doc, variable renames, and inject context

adisazhar123 committed:
test: assert storage id value after backfill

adisazhar123 committed:
test: assert storage id value on import

We removed the migrator mock and use our own fake implementation to assert the storage unique id value during import

adisazhar123 commented on upgrades/backend.go:
```diff
@@ -42,3 +56,175 @@ func (s stateBackend) AddVirtualHostKeys() error {
+			return nil, err
+		}
+
+		isLegacyDeployment := func(appName string) (bool, error) {
```
Done.

adisazhar123 commented on apiserver/facades/controller/migrationtarget/migrationtarget.go:
```diff
@@ -50,13 +77,46 @@ type APIV2 struct {
+
+// baseModel implements [stateenvirons.baseModel].
+type baseModel struct {
+	descriptionModel description.Model
```
Have created a func called `buildCloudSpec` to handle this. Unfortunately, description model doesn't have `cloud.Cloud` but it has cloud name, so we have to fetch `cloud.Cloud` by name from state. The rest are straight forward.

adisazhar123 commented on apiserver/facades/controller/migrationtarget/migrationtarget.go:
```diff
@@ -151,7 +213,59 @@ with an earlier version of the target controller and try again.
+			// We have detected an empty storage unique ID. It must be coming from
+			// an old model.
+			if k8sClient == nil {
+				k8s, err := api.initK8sClient(descriptionModel)
```
Done

adisazhar123 commented on apiserver/facades/controller/migrationtarget/migrationtarget.go:
```diff
@@ -162,6 +276,140 @@ func (api *API) Import(serialized params.SerializedModel) error {
+	}
+
+	deploymentName := app
+	isLegacy, err := isLegacyDeployment(app)
```
Done

adisazhar123 committed:
chore: reorder import

adisazhar123 committed:
feat: don't silent not found err from k8s

Let the caller continue the search if a not found err from k8s is returned

adisazhar123 committed:
feat: pass func to search for namespace

Makes it easier for unit testing. I also return early on the first storage id found.

adisazhar123 committed:
test: scenarios for getting storage ids from k8s

Tests cover the case where storage id exists in the annotation and when it doesn't it should generate one.

adisazhar123 committed:
chore: reorder import

wallyworld commented on apiserver/facades/controller/migrationtarget/migrationtarget.go:
```diff
@@ -150,8 +174,17 @@ with an earlier version of the target controller and try again.
+
 	controller := state.NewController(api.pool)
-	_, st, err := migration.ImportModel(controller, api.getClaimer, serialized.Bytes)
+	_, st, err := migration.ImportModel(controller, api.getClaimer, descriptionModel)
```
Is there a reason not to call `api.importModel` here too? I assume there's no test for this code path like there is for the caas case? Regardless, being consistent is better.

adisazhar123 replied:
I have standardized this to call `api.importModel`. It's much cleaner now.

The CAAS condition performs a backfill. The import for both CAAS and IAAS invokes the same func.

wallyworld commented on apiserver/facades/controller/migrationtarget/migrationtarget.go:
```diff
@@ -162,6 +195,131 @@ func (api *API) Import(serialized params.SerializedModel) error {
+	for _, findStorageUniqueID := range jujukubernetes.StorageUniqueIDFinder {
+		storageUniqueID, err := findStorageUniqueID(ctx, k8sClient, namespace,
+			app, modelName, modelUUID, controllerUUID)
+		if err != nil && k8serrors.IsNotFound(err) {
```
Don't need the `err != nil`

I'd also do the happy path first
```
if err ==nil {
    return storageUniqueID
}
if k8serrors.IsNotFound(err) {
    continue
}
return "", errors.Annotate(err, "finding storage unique ID")
```

wallyworld reviewed (commented):
Thanks for the changes. The migration stuff looks much cleaner.
I started commenting in the migration target code then looked art the tests and the test comments negate some of the others. Main issue is that we are setting up a full stack state test suite and then mocking out the bit that writes to state. To save refactoring the entire suite to remove state, we're better to be consistent and just run the import as we do for other tests and read the unique id from state to assert on it. Then we don't need to mock out the import func and the other churn that comes with it.

adisazhar123 committed:
refactor: introduce abstraction and funcs in constructor

I introduced MigrationState interface that represents a subset of state.State methods. In addition, there are two new func types, importModelFunc and precheckShimFunc. These new types makes it easier for mocking in unit tests.

adisazhar123 committed:
refactor: inject facade dependencies

I have created some wrapper functions as well so that they conform to the new types I have introduced.

adisazhar123 committed:
refactor: check happy condition before sad condition

adisazhar123 committed:
feat: generate new mocks

adisazhar123 committed:
test: adjust test to work with new types and move out new suite

adisazhar123 committed:
test: new suite without mongo dependency for caas migration

CodingCookieRookie commented on storage/kubernetes.go:
```diff
@@ -66,3 +74,12 @@ type KubernetesFilesystemUnitAttachmentParams struct {
 }
+
+// RandomPrefix returns a random string.
+func RandomPrefix() (string, error) {
```
hmm.. so technically all 3.6.12 and below applications will have a storage unique ID generated randomly using this?

CodingCookieRookie reviewed (approved):
LGTM, i did a qa test to migrate a model with mongodb-k8s from 3.6.11 to 3.6.12 controller and the storage unique id is populated as well, thanks for the great work!

wallyworld reviewed (commented):
We can simplify things, see the attached diff. We don't need the db model, just the state.
As a drive by, the migration tests in state we're closing the migrated state object.

https://pastebin.ubuntu.com/p/xdqSZNdMC2/

adisazhar123 committed:
refactor: remove unnecessary *state.Model

adisazhar123 committed:
test: adjust tests after removing *state.Model

adisazhar123 commented on upgrades/backend.go:
```diff
@@ -42,3 +55,96 @@ func (s stateBackend) AddVirtualHostKeys() error {
+						Name:            a.Name,
+						StorageUniqueID: storageUniqueID,
+					})
+					found = true
```
`k8sApps` contains all the apps we’ve already found storage IDs for. The `found` flag is used to check whether the current app was resolved. If not, we return an error right away.

Using `len(k8sApps) > 0` wouldn’t work, because it might just be holding previous apps. That could make it look like the current app was found when it actually wasn’t.

adisazhar123 committed:
refactor: rename upgrade target from 3.6.12 to 3.6.13

Renamed because this feature will be landed in 3.6.13. It didn't make the 3.6.12 release.

adisazhar123 commented:
QA passes after latest changes. :+1:

wallyworld commented on apiserver/facades/controller/migrationtarget/register.go:
```diff
@@ -82,5 +98,29 @@ func newFacade(ctx facade.Context, facadeVersions facades.FacadeVersions) (*API,
+
+// precheckShim wraps [migration.PrecheckShim] so we conform to the contract
+// in [NewAPI].
+func precheckShim(s *state.State) precheckShimFunc {
```
We're separating the "states" used to call precheck. We either want to pass in both model state and controller state here, or, better, just use a func with the same signature as `migration.Precheck` and in the facade code, pass in `api.state` and `controllerState`

wallyworld reviewed (approved):
Thank you.
I still think we don't need `found`, unless I'm missing something

adisazhar123 committed:
refactor: rename variable for missing k8s resource

adisazhar123 committed:
refactor: wrap precheckshim with model and controller state

adisazhar123 commented:
/merge
